decisions/0001-bundle-granularity.rst

```diff
@@ -0,0 +1,199 @@
+
+The ``static`` directory is optional. For a concrete example::
+
+    /problem+blockstore_granularity/
+                                    problem.xml
+                                    static/
+                                           diagram.png
+
+Meanwhile, a container looks like:
+
+    /unit+blockstore_big_questions/problem.xml
+
+When containers reference their children, it's done via a new tag. So a
+particular ``unit.xml`` might look like::
+
+    <unit display_name="">
+        <!-- Normal usage -->
+        <xblock-include definition="unit+blockstore_big_questions" />
+
+        <!-- If we want to include the same content with a modifier to the usage
+             key so that we can include it multiple times with different state
+             storage. -->
+        <xblock-include definition="unit+blockstore_big_questions" usage="alternate" />
+
+        <!-- XBlock that exists in a Linked Bundle aliased to "arch_hangout_vidoes" -->
+        <xblock-include definition="arch_hangout_videos/video+blockstore_overview"/>
+    </unit>
+
+Links
+=====
+
+Link information is stored at the Snapshot level. Since the human-readable
+format is going to be a transform, we don't have to make symlink analogies to
+make the data easier to work with in authoring. Instead, Links are a mapping of
+names to Bundle Snapshots::
+
+    {
+        "links": {
+            "arch_hangout_videos": {
+                "bundle_uuid": "3fcf5f61-bc23-41ec-9452-26d12dc3b13c",
+                "snapshot": "8f7bc6e89581591fa925fefa3819d382fd793839"
+                "dependencies": [
+                    {
+                        "bundle_uuid": "fd7ee4b3-0540-406c-98b9-dd050b7ddcb2",
+                        "snapshot": "c0c0940e4b3151908b60cecd1ef5e2aa19904676"
+                    }
+                ]
+            }
+        }
+    }
+
+Since all Snapshots capture their full dependencies, they can do cycle detection
+quickly. Since Links are explicitly named, we can keep multiple references to
+different versions of the same Bundle, which might be useful for things like
+shared code libraries.
+
+The notifications aspect of Bundles would be handled by a M:M mapping of Bundles
+(not BundleVersions), with a relationship of ``(lending_bundle, borrowing_bundle,
+still_using)``, where ``still_using`` is a boolean flag that is true if the most
+recent version of the borrowing Bundle still has a link to any version of the
+lending bundle.
+
+Changeset-Level Metadata
+========================
+
+Data about who changed what and why are stored at the BundleVersion level of
+granularity. This does mean that a Changelog for a Course would covered many
+different blocks, and that finding when a particular block changed would be
+more difficult. However, it gives a good high level view of the history of the
+work as a whole (the Course) in a simple way. In the Content Library case, each
+problem would be its own Bundle, so we'd still have tracking at the individual
```

Comment:
> This means that a course using a Problem Library with 10,000 problems still needs 10,000 Links to bundles, right?

If we're statically linking them in using the same mechanism that Content Libraries use today, that's true. I don't think we'd want that though. FWIW, the max I know of for problems in a single course is ~900.

> And how do we represent the fact that when additional bundles are added to the library, they need to be linked into the course?

Ack, that's sloppy of me. I think that there are really two kinds of Content Libraries we're talking about, and I need to 

1. The Content Library of today, where the contents are statically copied into a Sequence at publish time, should be its own Bundle, with each problem being a directory.
2. The Problem Bank of the FUTURE, where each problem is a Bundle and we'll pulling things back more dynamically. The Problem Bank as a whole would be represented as a Collection, and Courses wouldn't have a hardcoded reference to each Bundle separately.

Reply:
> The Problem Bank as a whole would be represented as a Collection, and Courses wouldn't have a hardcoded reference to each Bundle separately.

But we still need to know which courses are using which libraries to facilitate update notifications. So if a Library = Collection, then that means courses would need to be able to link to Collections as well as to Bundles?

Why not just keep it simple and say 1 Course = 1 Bundle, _and_ 1 Library = 1 Bundle ?

Reply:
I think it's because I have this vague sense that grouping the problems together will make less sense as content libraries grow larger, and that it will be less necessary to send those types of notifications as we shift further away from courses. If I have a course and I care that you learn about fast fourier transforms, I feel like at some point we'll cross a bridge from "I'm going to import and use these nifty FFT problems in my course" to "I'm going to let the adaptive system teach you FFT, and then let you come back to my course so that I can teach you the thing that builds on top of that."

That being said, I think that it's a safe default going in position to say 1 Course = 1 Bundle and 1 Library = 1 Bundle, so long as there is a straightforward mapping out of that -- which I think we have as long as the problems are cleanly separable from each other. I'll update the proposal to reflect that.